### PR TITLE
Add prognostic cumulus closure description in saSAS

### PR DESCRIPTION
**Pull request by lisa-bengtsson**
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [ ] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsiblity to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

This PR introduces an optional closure assumption to the Arakawa-Schubert cloud work function in saSAS cumulus convection, in where the mass-flux at cloud base is described as the product between updraft area fraction and updraft velocity. The updraft area fraction is given by a prognostic moisture budget equation.

The update is not aimed for UFS coupled prototype 8, but is an option to be explored beyond P8.

In addition to updates in the ccpp/physics submodule, new field_table, and namelist parameters are added in ufs-weather-model.

The PR is also making a small change in the stochastic_physics_driver.F90 to allow the cellular automata to work on non-uniform blocksizes. Previously the code aborted if not all blocks in the domain had the same size, however, the code works the way it's been formulated, so I simply removed the call to mpp_error in case on non-uniform blocksizes. 

##Testing

This PR does not change the answers in the control, but a new test has been added where the scheme is activated using progsigma = True.

The regression tests have been tested on Hera.

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [x] wcoss2
- [x] opnReqTest for newly added/changed feature
- [ ] CI

## Dependencies

- waiting on https://github.com/NCAR/ccpp-physics/pull/903
- waiting on https://github.com/NOAA-EMC/fv3atm/pull/523

Issue: https://github.com/NCAR/ccpp-physics/issues/890
https://github.com/ufs-community/ufs-weather-model/issues/1193

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,15 +1,11 @@
 [submodule "FV3"]
   path = FV3
-  url = https://github.com/NOAA-EMC/fv3atm
+  url = https://github.com/lisa-bengtsson/fv3atm
   branch = develop
```
Comment (grantfirl) [inline .gitmodules]: The branch should also point to your prog_closure branch instead of develop here.

File: tests/rt.conf [hunk 2]
```diff
@@ -19,6 +19,7 @@ RUN     | cpld_restart_c192_p8
 
 RUN     | cpld_control_c384_p8                                                                                                    | - wcoss_cray wcoss2.intel jet.intel     | fv3 |
 RUN     | cpld_restart_c384_p8                                                                                                    | - wcoss_cray wcoss2.intel jet.intel     |     | cpld_control_c384_p8
+RUN     | cpld_control_c384_p8_progsigma                                                                                          | - wcoss_cray wcoss2.intel jet.intel     | fv3 |
```
Comment (lisa-bengtsson) [inline tests/rt.conf]: @junwang-noaa I am not sure if this test should only be tested on  "wcoss_cray, wcoss2.intel, jet.intel" as listed (same as cpld_control_c384_p8). Let me know what you think. For me this is OK.

**Comment by DeniseWorthen** [comment]
@lisa-bengtsson The cpld_control_c384 test will be deleted when we commit the mesh cap PR and the bmark_p8 (+ restart) will be the only c384 coupled test we carry forward in the regression tests.

**Comment by lisa-bengtsson** [comment]
> @lisa-bengtsson The cpld_control_c384 test will be deleted when we commit the mesh cap PR and the bmark_p8 (+ restart) will be the only c384 coupled test we carry forward in the regression tests.

Ok @DeniseWorthen , would it be better to add a bmark_p8_c384_progsigma test?

**Comment by DeniseWorthen** [comment]
Could you add this feature test at C96-1deg?

**Comment by lisa-bengtsson** [comment]
> Could you add this feature test at C96-1deg?

Unfortunately not, the closure is scale adaptive and does not have an impact at such coarse resolution. It could be an atmosphere only test (e.g. control_c384_progsigma), but I did not see any other control_c384 tests, so not sure how create IC's for such a test.

**Comment by DeniseWorthen** [comment]
OK, thanks. We'll discuss this at tomorrow's morning tag-up.

**Comment by lisa-bengtsson** [comment]
Thank you

**Comment by DeniseWorthen** [comment]
@lisa-bengtsson Could you create this feature test for the regular RTs from the control_c384 standalone test? 

After the mesh cap PR, we will create a bmark_p8 feature test and add it to the weekly tests. Those run for 5 days but do not compare against a baseline. They are intended more as basic sanity/stability tests for the coupled prototype configuration.

**Comment by lisa-bengtsson** [comment]
hm, ok, I had missed that there was a control_c384 test, yes that is better, I will update the test to a control_c384_prgosgima test instead.

**Comment by lisa-bengtsson** [comment]
Now I recall why I didn't add it to control_c384, I was looking for a control_**p8**_c384 as I was told all the "control" tests are planned to be replaced by "control_p8". But perhaps it is still OK to add a test on top of control, even if the plan is to replace them all?

**Comment by DeniseWorthen** [comment]
The only standalone p8 tests we currently have are low-resolution (c96) without aerosols. Yes, you could create a feature test from the control_c384 and it would get updated at some point to a P8 configuration. Is that sufficient or do you really need to add this to a P8 configuration in order for it to be a valid test?

**Comment by lisa-bengtsson** [comment]
The control_c384 is sufficient since it also uses the convection scheme which I have updated in this PR. I will run some tests and update my PR later today.

**Comment by DeniseWorthen** [comment]
That's great. Thanks.

File: tests/parm/control_progsigma.nml.IN [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  lw_file_clouds     = 'rrtmgp-cloud-optics-coeffs-lw.nc'
+  sw_file_gas        = 'rrtmgp-data-sw-g112-210809.nc'
+  sw_file_clouds     = 'rrtmgp-cloud-optics-coeffs-sw.nc'
+  rrtmgp_nGptsSW     = 112
+  rrtmgp_nGptsLW     = 128
+  rrtmgp_nBandsLW    = 16
+  rrtmgp_nBandsSW    = 14
+  doGP_cldoptics_LUT = @[DOGP_CLDOPTICS_LUT]
+  doGP_lwscat        = @[DOGP_LWSCAT]
+  progsigma          = .true.
```
Comment (DeniseWorthen) [inline tests/parm/control_progsigma.nml.IN]: The file control_prosigma.nml.IN looks identical to control.nml.IN with the addition of the prosigma variable. Please instead add a new variable in default_vars (export PROSIGMA=.false.) and then set it true in your control_c384_prosigma test. You would not need to add the control_prosigma.nml.IN.

Comment (lisa-bengtsson) [reply tests/parm/control_progsigma.nml.IN]: Ok, sure I can do that.

Comment (jkbk2004) [reply tests/parm/control_progsigma.nml.IN]: @lisa-bengtsson so is the PR ready to run regression test? I mean where to use PROSIGMA option in control_prosigma or control_c384_prosigma?

Comment (lisa-bengtsson) [reply tests/parm/control_progsigma.nml.IN]: Running some tests now, I will push these updates in about an hour, is this PR still planned to be merged tomorrow?

Comment (jkbk2004) [reply tests/parm/control_progsigma.nml.IN]: @lisa-bengtsson cheyenne.intel is last RT test going on #1218. Then, we can start working this PR. 6/2 commit date for this PR is still good.

Comment (lisa-bengtsson) [reply tests/parm/control_progsigma.nml.IN]: @DeniseWorthen @junwang-noaa @jkbk2004 I would like to commit one more update before merging, the c384 dcp ORT does not work (control_c384), @MinsukJi-NOAA has provided a potential solution that we are testing right now. It takes a bit of time in queue with the c384 tests. If Minsuk's solution works then I would like to add it to this PR.

Comment (jkbk2004) [reply tests/parm/control_progsigma.nml.IN]: @lisa-bengtsson sure!

Comment (lisa-bengtsson) [reply tests/parm/control_progsigma.nml.IN]: The tests

./opnReqTest_l -n control_c384 -c dcp 
./opnReqTest_l -n control_c384_progsigma -c dcp

Both run and passes with Minsuk's change 

Now we are testing:

./opnReqTest_l -n control_p8 -c dcp

Just to be sure that still works. If this works I will push Minsuk's update in opnReqTests/dcp.sh to allow ORTs to work with c384 resolution.

**Comment by lisa-bengtsson** [comment]
I have pushed the updates to address Denises's comments regarding my newly added test, and Minsuk's ORT test fix. The PR is ready on my end.

**Comment by MinsukJi-NOAA** [comment]
ORT change looks good. Thanks for testing @lisa-bengtsson

File: .gitmodules [hunk 4]
```diff
@@ -42,3 +34,9 @@
   path = AQM
   url = https://github.com/NOAA-EMC/AQM
   branch = develop
+[submodule "stochastic_physics"]
+	path = stochastic_physics
+	url = https://github.com/noaa-psd/stochastic_physics
+[submodule "CDEPS-interface/CDEPS"]
```
Comment (junwang-noaa) [inline .gitmodules]: May I ask why you make changes in the stochastic_physucs and CDEPS submodules?

Comment (lisa-bengtsson) [reply .gitmodules]: I think .gitmodules got a bit re-shuffled after the stochastic_physics fiasco where I removed and added the submodule. Typing 'git submodule add' put the stochastic_physics submodule path at the end. I have cleaned it up now.

**Comment by jkbk2004** [comment]
@junwang-noaa Can I confirm that we will set BL_DATE=20220602 to create new baseline, right?

**Comment by jkbk2004** [comment]
@lisa-bengtsson @junwang-noaa sounds like creating single BL for control_c384_progsigma and copying over to current will be quicker. @DeniseWorthen I will try to create one on cheyenne first.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: jet
Compiler: intel
Job: RT
[RT] Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/913741286/20220602174517/ufs-weather-model
[RT] Error: Test control_c192 017 failed in run_test failed 
Please make changes and add the following label back: jet-intel-RT

**Comment by jkbk2004** [comment]
@DeniseWorthen @MinsukJi-NOAA @junwang-noaa @grantfirl RT passes ok! CI is running. Can you leave review comments?

**Comment by grantfirl** [comment]
@jkbk2004 @junwang-noaa I'm at a conference today, but I can start the merge process when ready if before 12 ET. After 12 ET, I'll be traveling home for the remainder of the afternoon. If necessary, I can also delegate this to someone else.

**Review by junwang-noaa — APPROVED** [review]

**Review by grantfirl — APPROVED** [review]

**Comment by jkbk2004** [comment]
@grantfirl can we start merging in [ccpp#903](https://github.com/NCAR/ccpp-physics/pull/903) and then [fv3atm#523](https://github.com/NOAA-EMC/fv3atm/pull/523) ?

**Comment by grantfirl** [comment]
> @grantfirl can we start merging in [ccpp#903](https://github.com/NCAR/ccpp-physics/pull/903) and then [fv3atm#523](https://github.com/NOAA-EMC/fv3atm/pull/523) ?

I've merged https://github.com/NCAR/ccpp-physics/pull/903 and instructed @lisa-bengtsson to make the changes to her fv3atm branch in anticipation of the merge.

**Comment by jkbk2004** [comment]
@lisa-bengtsson can you revert the fv3atm pointer here as well? so we can merge in.

**Comment by lisa-bengtsson** [comment]
@jkbk2004 I updated the .gitmodules and FV3

**Comment by lisa-bengtsson** [comment]
Thank you all for all the work on running RT's and merging this!

**Pull request merged**